rand_chacha/src/chacha.rs

```diff
@@ -241,6 +241,14 @@ macro_rules! chacha_impl {
                 }
             }
         }
+
+        impl PartialEq<$ChaChaXRng> for $ChaChaXRng {
+            fn eq(&self, rhs: &$ChaChaXRng) -> bool {
+                self.rng.core.state.stream64_eq(&rhs.rng.core.state)
+                    && self.get_word_pos() == rhs.get_word_pos()
```

Comment:
This should be correct, assuming we can trust that the contents of the buffer are equal, which we should be able to considering this is cleared any time the stream position is changed.

*But* I notice that `get_word_pos` assumes we have a 16-word buffer. We have a 64-word buffer now. Whoops (I guess we should multiply by the buffer length instead of a fixed left-shift).

Fixing this now of course has other consequences and is very much a breaking change (though we're landing breaking changes anyway). Can you add a note to the changelog?

Reply:
> This should be correct, assuming we can trust that the contents of the buffer are equal, which we should be able to considering this is cleared any time the stream position is changed.

Buffer content is definitely made redundant by `ChaCha` state. For any `ChaCha` state, there is a unique corresponding state of the `BlockRng` buffer such that if the observable portion of the buffer were not consistent with this state, it would be possible to observe between successive calls to `get_u64()` a change in effective parameters outside of the normal block increment. That would reflect a bug in a mutating method.

> *But* I notice that `get_word_pos` assumes we have a 16-word buffer. We have a 64-word buffer now. Whoops (I guess we should multiply by the buffer length instead of a fixed left-shift).

The left-shift is converting from a value measured in ChaCha blocks (the size of which is fixed by the algorithm definition) to a value measured in words. It's correct, just hard to read. I'll fix that.

The other line `block -= 4` does rely on the buffer size remaining 64 words (4 blocks) though, which is fragile.

While I was making `get_word_pos` a bit more readable I did notice bugs in [handling the counter-wrap edge case](https://github.com/rust-random/rand/pull/980).

Reply:
Good, I believe this is resolved.

Can you rebase or merge this PR please?